rosbag2_cpp/src/rosbag2_cpp/info.cpp

```diff
@@ -127,20 +171,112 @@ std::vector<std::shared_ptr<rosbag2_service_info_t>> Info::read_service_info(
+              }
+              break;
+            }
+          case service_msgs::msg::ServiceEventInfo::RESPONSE_SENT:
+          case service_msgs::msg::ServiceEventInfo::RESPONSE_RECEIVED:
+            {
+              if (action_topic_type == TopicsInAction::SendGoalEvent) {
+                action_process_info[action_name]->
+                send_goal_service.response[msg.client_gid].emplace(msg.sequence_number);
+              } else if (action_topic_type == TopicsInAction::GetResultEvent) {
+                action_process_info[action_name]->
+                get_result_service.response[msg.client_gid].emplace(msg.sequence_number);
+              } else {
+                // TopicsInAction::CancelGoalEvent
+                action_process_info[action_name]->
+                cancel_goal_service.response[msg.client_gid].emplace(msg.sequence_number);
+              }
+              break;
+            }
+        }
+      } else {
+        // Handle service event topic
+        switch (msg.event_type) {
+          case service_msgs::msg::ServiceEventInfo::REQUEST_SENT:
+          case service_msgs::msg::ServiceEventInfo::REQUEST_RECEIVED:
+            service_process_info[bag_msg->topic_name]->request[msg.client_gid].emplace(
+              msg.sequence_number);
+            break;
+          case service_msgs::msg::ServiceEventInfo::RESPONSE_SENT:
+          case service_msgs::msg::ServiceEventInfo::RESPONSE_RECEIVED:
+            service_process_info[bag_msg->topic_name]->response[msg.client_gid].emplace(
+              msg.sequence_number);
+            break;
+        }
+      }
+    }
+
+    // Process action_process_info to get the number of request and response
+    for (auto & [action_name, action_info] : action_process_info) {
+      size_t count = 0;
+      // Get the number of request from all clients for send_goal
+      for (auto &[client_id, request_list] : action_info->send_goal_service.request) {
+        count += request_list.size();
+      }
+      all_action_info[action_name]->send_goal_service_msg_count.first = count;
+
+      // Get the number of request from all clients for cancel_goal
+      count = 0;
+      for (auto &[client_id, request_list] : action_info->cancel_goal_service.request) {
+        count += request_list.size();
+      }
+      all_action_info[action_name]->cancel_goal_service_msg_count.first = count;
+
+      // Get the number of request from all clients for get_result
+      count = 0;
+      for (auto &[client_id, request_list] : action_info->get_result_service.request) {
+        count += request_list.size();
+      }
+      all_action_info[action_name]->get_result_service_msg_count.first = count;
+
+      // Get the number of response from all clients for send_goal
+      count = 0;
+      for (auto &[client_id, response_list] : action_info->send_goal_service.response) {
+        count += response_list.size();
+      }
+      all_action_info[action_name]->send_goal_service_msg_count.second = count;
```

Comment:
Please consider moving these data processing blocks to a separate function. i.e. for the 
`// Process action_process_info to get the number of requests and responses` and `// Get the number of response from all clients for get_result`